app/lib/domains.server.ts

```diff
@@ -0,0 +1,138 @@
+import * as dns from './dns.server';
+import { RecordStatus } from '@prisma/client';
+import { prisma } from '~/db.server';
+import type { RecordType, Record } from '@prisma/client';
+
+interface DnsRecord {
+  type: RecordType;
+  name: string;
+  value: string;
+}
+
+interface DomainRecord {
+  id?: Record['id'];
+  username?: Record['username'];
+  description?: Record['description'];
+  course?: Record['course'];
+  ports?: Record['ports'];
+}
+
+type Domain = DnsRecord & DomainRecord;
+
+export const createUserDomain = async (data: Domain) => {
+  if (!data.username) {
+    throw new Error('Username is not provided');
+  }
+
+  const existingRecords = await prisma.record.count({
+    where: {
+      name: data.name,
+      type: data.type,
+      value: data.value,
+    },
+  });
+
+  if (existingRecords > 0) {
+    throw new Error('Record already exists');
+  }
+
+  await dns.createRecord(data.type, data.name, data.value);
+
+  const result = await prisma.record.create({
+    data: {
+      username: data.username,
+      type: data.type,
+      name: data.name,
+      value: data.value,
+      expiresAt: setMonthsFromNow(6),
+      status: RecordStatus.pending,
+    },
+  });
+
+  if (!result) {
+    throw new Error('Could not create a record in DB');
+  }
+  return result;
+};
+
+export const updateUserDomain = async (data: Domain) => {
+  await dns.upsertRecord(data.type, data.name, data.value);
```

Comment:
Same pattern here, and probably everywhere below: you can run the db + dns async operations at the same time using `Promise.all()`